lib/cadet/assessments/assessments.ex

```diff
@@ -481,6 +506,112 @@ defmodule Cadet.Assessments do
     Question.changeset(%Question{}, params_with_assessment_id)
   end
 
+  def insert_voting(
+        contest_number,
+        usernames,
+        question_id
+      ) do
+    contest_assessment = Repo.get_by(Assessment, number: contest_number)
+    users = Enum.map(usernames, fn x -> Repo.get_by(User, username: x.username) end)
+    invalid_usernames = Enum.any?(users, fn user -> user == nil end)
+
+    if contest_assessment != nil and !invalid_usernames do
+      contest_submission_ids =
+        Submission
+        |> where(assessment_id: ^contest_assessment.id)
+        |> where(status: "submitted")
+        |> Repo.all()
+        |> Enum.map(fn x -> x.id end)
+
+      user_ids = Enum.map(usernames, fn x -> Repo.get_by(User, username: x.username).id end)
+
+      votes_per_user = min(length(contest_submission_ids), 10)
+
+      votes_per_submission =
+        if Enum.empty?(contest_submission_ids) do
+          0
+        else
+          trunc(Float.ceil(votes_per_user * length(usernames) / length(contest_submission_ids)))
+        end
+
+      submission_id_map =
+        contest_submission_ids
+        |> Enum.into(%{}, fn id -> {id, votes_per_submission} end)
+
+      {_submission_map, submission_votes_changesets} =
+        user_ids
+        |> Enum.reduce({submission_id_map, []}, fn user_id, acc ->
+          {submission_map, submission_votes} = acc
+
+          user_contest_submission =
+            Repo.get_by(Submission, student_id: user_id, assessment_id: contest_assessment.id)
+
+          {user_submission_id_votes_left, submission_map} =
+            if is_nil(user_contest_submission) do
+              {nil, submission_map}
+            else
+              user_submission_id_votes_left = Map.get(submission_map, user_contest_submission.id)
+              submission_map = Map.delete(submission_map, user_contest_submission.id)
+              {user_submission_id_votes_left, submission_map}
+            end
+
+          submission_ids = Enum.take_random(submission_map, votes_per_user)
```

Comment:
This line is O(n), and this whole reduce is O(m) where n is # submissions and m is # users.. Not gonna have a good time when we have 600+ users and submissions. (Or more.)

You might try taking the list of submission IDs and multiplying it by the number of votes per submission. Then shuffle that entire list once. Then you can just take the first N elements, and replace if there are duplicates or self-votes.